db/sequence_allocator.go

```diff
@@ -381,3 +415,69 @@ func (s *sequenceAllocator) waitForReleasedSequences(ctx context.Context, startT
 	time.Sleep(requiredWait)
 	return requiredWait
 }
+
+// _fixSyncSeqRollback will correct a rolled back _sync:seq document in the bucket
+func (s *sequenceAllocator) _fixSyncSeqRollback(ctx context.Context, incrValue, expectedValue uint64) (allocatedToSeq uint64, err error) {
+	// grab _sync:seq value + its current cas value
+	var result uint64
+	cas, err := s.datastore.Get(s.metaKeys.SyncSeqKey(), &result)
+	if err != nil {
+		return 0, err
+	}
+	base.DebugfCtx(ctx, base.KeyCRUD, "correcting _sync:seq document. Incrementing _sync:seq document from %d by the value of %d", result, incrValue)
+
+	worker := func() (bool, error, interface{}) {
+		// set the value to _sync:seq current value + incr value
+		setVal := result + incrValue
+		_, err = s.datastore.WriteCas(s.metaKeys.SyncSeqKey(), 0, cas, setVal, 0)
+		if err == nil {
+			return false, nil, setVal
+		}
+		// bail out if we don't get CAS error
+		if !base.IsCasMismatch(err) {
+			return false, err, 0
+		}
+		// fetch _sync:seq doc to update cas + check if it has already been fixed by another process
+		var newResult uint64
+		cas, err = s.datastore.Get(s.metaKeys.SyncSeqKey(), &newResult)
+		if err != nil {
+			return false, err, nil
+		}
+		shouldRetry := shouldRetrySyncSeqCorrection(setVal, newResult)
+		if shouldRetry {
+			// update result value of the new value fetched above before retrying
+			result = newResult
+			return true, err, nil
+		}
+		// if no retry is needed (_sync:seq has potentially been corrected by another node),
+		// we still need to increment by batch size to get unique batch for this node
+		newResult, err = s.incrementSequence(s.sequenceBatchSize)
+		if err != nil {
+			return false, err, 0
+		}
+		return false, err, newResult
+	}
+
+	// in case _sync:seq was fixed in the time between detecting rollback and fetching cas from the doc
+	if result < expectedValue {
+		err, currSeq := base.RetryLoop(ctx, "Fix _sync:seq Value", worker, base.CreateDoublingSleeperFunc(1000, 5))
+		if err != nil {
+			return 0, err
+		}
+		allocatedToSeq = currSeq.(uint64)
+	} else {
+		// if _sync:seq has been fixed just increment by batch size to gte new batch for this node
+		base.DebugfCtx(ctx, base.KeyCRUD, "_sync:seq document fixed by another process, continuing with sequence batch allocation")
+		allocatedToSeq, err = s.incrementSequence(s.sequenceBatchSize)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	return allocatedToSeq, err
+}
+
+// shouldRetrySyncSeqCorrection returns true if value of _sync:seq is less than the expected value, else returns false
+func shouldRetrySyncSeqCorrection(expectedVal, actualVal uint64) bool {
```

Comment:
I think a separate function for "actual < expected" is overkill and makes the calling code harder to read.